dist_run.py

```diff
@@ -235,31 +337,179 @@ def main():
+    )
+    logger.info(f"{prompt_lengths=}")
+    logger.info(f"first prompt {padded_sequence[0, :prompt_lengths[0]+1]=}")
+    if len(prompt_lengths) > 1:
+        logger.info(f"second prompt {padded_sequence[1, :prompt_lengths[1]+1]=}")
 
     schedule = ScheduleGPipe(stage, mbs)
     logger.info(f"Created schedule: {schedule}")
 
-    with torch.no_grad():  # .inference_mode():
-        if pp_rank == 0:
-            schedule.step(input_ids)
-        else:
-            output = schedule.step()
+    # with CUDATrackTime() as timer:
+    first_pp_group = 0
+    last_pp_group = pp_group_size - 1
+
+    x_recv = torch.zeros(1, device=device, dtype=torch.int64)
+    logger.info(f"{x_recv.shape=}")
+
+    last_global_rank = world_size - 1
+    res = []
+    dst = None
+    src = None
+
+    if pp_rank == last_pp_group:
+        dst = dist.get_global_rank(pp_group, 0)
+    elif pp_rank == 0:
+        src = dist.get_global_rank(pp_group, last_pp_group)
 
+    # Decoding
+    num_tokens = 10
+    """
+    with torch.no_grad():
+        for step in range(num_tokens + 1):  # +1 to include the initial prefill step
+            if pp_rank == 0:
+                schedule.step(padded_sequence)
+                dist.recv(x_recv, src, group=pp_group)
+                logger.info(f"RECEIVED {x_recv=}")
+                assert x_recv != 128006, f"next_token is header id={x_recv}"
+                _update_padded_sequence(padded_sequence, x_recv, res, prompt_lengths)
+                logger.info(
+                    f"Updated padded seq start: {prompt_lengths=}, {padded_sequence[:, prompt_lengths[0]-1]=}"
+                )
+
+            elif pp_rank == last_pp_group:
+                output = schedule.step()
+                decode_results = _batch_decode_next_tokens(
+                    output, prompt_lengths, tokenizer
+                )
+                logger.info(
+                    f"\n\n{color.green} {'Prefill' if step == 0 else '* Decode *'} responses ====>>>> {color.blue} {decode_results=} \n{color.reset}"
+                )
+
+                next_token = torch.tensor([decode_results[0][0]], device=device)
+                res.append(decode_results[0][1])
+
+                # increment prompt lengths for next token
+                for i in range(len(prompt_lengths)):
+                    prompt_lengths[i] += 1
+                    logger.info(
+                        f"output review {prompt_lengths[i]=}, {padded_sequence[i, prompt_lengths[i]-1]=}"
+                    )
+
+                if step < num_tokens - 1:
+                    dist.send(next_token, dst, pp_group)
+                logger.info(
+                    f"SENDING back...from rank={pp_rank} to dst={dst}, data {next_token.shape=}, {next_token=}"
+                )
+                assert next_token != 128006, f"next_token is header id={next_token}"
+
+            else:  # middle pp ranks
+                schedule.step()
+    """
```

Comment:
do you mind moving commented-out block to a different PR if not required by this PR?